File: backend/src/main/java/com/own/expertfinder/model/Customer.java
package com.own.expertfinder.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.own.expertfinder.interfaces.GeneralUserInterface;
import com.own.expertfinder.util.HashMapConverter;

import javax.persistence.*;
import java.io.Serializable;
import java.util.Date;
import java.util.Map;

@Entity
@Table(name = "customers", schema = "public")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class Customer extends AbstractModel implements Serializable, GeneralUserInterface {
    private String email;
    @OneToOne(cascade = CascadeType.ALL, fetch = FetchType.EAGER)
    @JoinColumn(name = "user_id")
    private User user;
    private String firstName;
    private String lastName;
    private String phoneNumber;
    private String profession;
    private Boolean isActive;
    // TODO Choose the right solution
    // @Convert(converter = HashMapConverter.class)
    // private Map<String, Object> position;
    private String position;
    /**
     * - 0 -> Available
     * - 1 -> Not available
     */
    private Integer status;
    // private Integer availableFrom; // TODO type?
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private Date registrationDate;


    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    public String getProfession() {
        return profession;
    }

    public void setProfession(String profession) {
        this.profession = profession;
    }

    public Boolean getActive() {
        return isActive;
    }

    public void setActive(Boolean active) {
        isActive = active;
    }

    public String getPosition() {
        return position;
    }

    public void setPosition(String position) {
        this.position = position;
    }

    public Integer getStatus() {
        return status;
    }

    public void setStatus(Integer status) {
        this.status = status;
    }

   /*
    public Integer getAvailableFrom() {
        return availableFrom;
    }

    public void setAvailableFrom(Integer availableFrom) {
        this.availableFrom = availableFrom;
    }
     */

    public Date getRegistrationDate() {
        return registrationDate;
    }

    public void setRegistrationDate(Date registrationDate) {
        this.registrationDate = registrationDate;
    }

}

File: frontend/src/app/models/customer.ts
import { User } from './user';
import { Status } from './enums';
import { Position } from './interfaces';

// Professional
export class Customer extends User {
  // tslint:disable:variable-name
  private _firstName: string;
  private _lastName: string;
  private _phoneNumber: string;
  private _professionId: number;
  private _isActive: boolean;
  // '_position' arrives from database as a 'string'
  // then it will be a 'Position' object
  private _position: Position | string;
  private _status: Status | null;
  private _availableFrom: Date; // ?
  private _user: User;

  constructor() {
    super();
  }

  public get firstName(): string {
    return this._firstName;
  }

  public set firstName(firstName: string) {
    this._firstName = firstName;
  }

  public get lastName(): string {
    return this._lastName;
  }

  public set lastName(lastName: string) {
    this._lastName = lastName;
  }

  public get phoneNumber(): string {
    return this._phoneNumber;
  }

  public set phoneNumber(phoneNumber: string) {
    this._phoneNumber = phoneNumber;
  }

  public get professionId(): number {
    return this._professionId;
  }

  public set professionId(profession: number) {
    this._professionId = profession;
  }

  public get isActive(): boolean {
    return this._isActive;
  }

  public set isActive(isActive: boolean) {
    this._isActive = isActive;
  }

  public get position(): Position | string {
    return this._position;
  }

  public set position(position: Position | string) {
    this._position = position;
  }

  public get status(): Status | null {
    return this._status;
  }

  public set status(status: Status) {
    this._status = status;
  }

  public get availableFrom(): Date {
    return this._availableFrom;
  }

  public set availableFrom(time: Date) {
    this._availableFrom = time;
  }

  public get user(): User {
    return this._user;
  }

  public set user(user: User) {
    this._user = user;
  }



}
